### ISSUE TITLE
ale should override signs from gitgutter

### ISSUE THREAD

**Issue by codeHatcher** [dense-analysis/ale#569]
Ale signs should take precedence over gitgutter, that's also how syntastic and gitgutter work together with syntastic taking precedence. the idea being as I fix my syntax errors, I'll be able to see my gitgutter signs. Especially important because I should be fixing syntax errors before committing code (which would clear the gitgutter signs)

Here is a related discussion on this for gitgutter and syntastic quite some time ago https://github.com/airblade/vim-gitgutter/issues/1

**Comment by w0rp** [comment]
This follows on from #23, one of the earliest issues, where I made it so it doesn't remove gitgutter signs. This might be slightly harder to implement than for Syntastic, as ALE updates signs as you type. If anyone wants to try fixing this, give it a go.

**Comment by codeHatcher** [comment]
Yup, I noticed that as well. Originally I was using Syntastic which is where I got use to the functionality. I happened to reinstall my dev environment and only now was wondering where the linter went. For now I can disable gitgutter until someone is able to come up with a fix or I move to Syntastic/Neomake.

@w0rp is it something that may not be possible with Ale? Helpful to know for deciding how long I should wait to switch over. Is there a priority level for what displays in the sign column?

**Comment by w0rp** [comment]
I reckon it should be possible. Someone just needs to go ahead and do it.

**Comment by codeHatcher** [comment]
@w0rp do you have a bounty bot you use, I wouldn't mind attaching a couple bucks to this issue

**Comment by codeHatcher** [comment]
I'm wrong, my issue was because I was using emoji for the gutter for ale and I see there is another issue here https://github.com/w0rp/ale/issues/511 that references a problem with tmux 2.4. In order to get both my gitgutter and ale signs to display, I changed the ale sign to simple characters and now everything looks ok

![image](https://cloud.githubusercontent.com/assets/5360982/26287031/6988de4a-3e40-11e7-8abf-d5abb806c96f.png)

**Comment by w0rp** [comment]
Cool. :+1:

**Comment by tjkirch** [comment]
I still have this issue when using the default ">>" and "--" ASCII sign characters for ale.  Same thing if they're single characters, e.g. ">" and "-".  gitgutter overrides ale, but I definitely want errors to take priority over git status.

My workaround is to have ale show the location list on error, but I'd prefer the sign characters alone with the ability to open the location list when needed.

I'm using vim 8.0.1415, ale 1568bf8128, and vim-gitgutter 1e60667322, all the latest as of now, and no tmux.

**Comment by w0rp** [comment]
As far as I can tell, whichever sign is set last wins. If vim-gitgutter sets signs after ALE, the signs will cover ALE's signs. Someone might have to change `vim-gitgutter` so it doesn't cover existing signs.

**Comment by unode** [comment]
Unsure if this is the same issue or a different one. If not I'll move this to another issue.

If I use the default `>>` and `--` I see two columns of signs both with a yellow background. By setting them to single characters `>` and `-` I expected only one column to be shown. However 2 are still visible, the second being just the yellow background with a single space (`> ` or `- `). I checked the source but didn't find any obvious way to customise this. Is this something that can be corrected or even adjusted automatically depending on the length of `ale_sign_error` and `ale_sign_warning`?

**Comment by w0rp** [comment]
The sign column is fixed to two characters in size in Vim, and cannot be changed.

**Comment by unode** [comment]
Thanks for the clarification.

**Comment by hsanson** [comment]
Not sure how this was working with syntactic+gitgutter, maybe luck?  The problem of sign conflicts between plugins is known from some time already and there is an issue discussion to improve the signs API (e.g. add priorities) in Neovim: https://github.com/neovim/neovim/issues/8668

**Comment by da-x** [comment]
In Neovim (and possibly in Vim, haven't looked at source), what currently determines the displayed sign, is the rather random order of link list of signs, order that is determined by the order of insertion. The one that appears first in the list is displayed.

I am working on a discussed change in Neovim where multiple sign column are displayed: https://github.com/neovim/neovim/pull/9295, and among it was agreed to sort by sign Id, so that more than one of them are displayed and the order is deterministic.

![image](https://user-images.githubusercontent.com/321273/49648814-e5c37900-fa2f-11e8-8d92-8841b54ef8c5.png)

Still in RFC stage.

**Comment by hsanson** [comment]
With #2786 it is possible to set g:ale_sign_priority to a value higher than 10 to ensure ALE signs take precedence over other plugins. Also this PR adds support for sign groups that ensure ALE sign manipulations do not interfere with other plugins signs.

Notes:
  1. Only works for NeoVim 0.4.0 and Vim 8.1.0614  and latter.
  2. Other plugins may also support sign-priority. Ensure you set g:ale_sign_priority to your needs.
      gitgutter does not use priority therefore any value larger than 10 ensures ALE will take 
      precedence over it.

**Comment by w0rp** [comment]
Thanks for your work @hsanson. :+1:

**Comment by IAXES** [comment]
@hsanson Good day. Does #2786 also handle use cases where VIM's built-in Git gutter signs (i.e. `+`, `-`, `=`) override Ale? Right now, using the latest VIM source available as of today, and the latest available (via Vundle) release of `w0rp/ale`, Python linting, for example, is completely broken, unless I copy my source out of the Git project (i.e. Gutter is completely disabled wrt Ale).

If I add the `'tomtom/quickfixsigns_vim'` plugin to the mix, VIM's built-in Git gutter functionality is working again, but Ale-related gutter functionality is completely overriden by the git feature. I'm aiming to either get both of these features (git gutter feature built-in to VIM, Ale) going with Ale overriding VIM/git gutter functionality, or I might just re-compile VIM without this Git gutter functionality if need be.

**Comment by vanphandinh** [comment]
@hsanson 
I'm using VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Jan  8 2020 00:57:57)
But It seems "sign-priority" not working..

I have added:
let g:gitgutter_sign_priority = 10
let g:ale_sign_priority = 11

But gitgutter's signs always show.

**Comment by cridemichel** [comment]
Hi all,
a possible workaround to obtain the correct behavior is to set the gitgutter priority
to a value less than 10.
Indeed, if in my .vimrc I set 
```vim
let g:ale_sign_priority=30 
let g:gitgutter_sign_priority=10
```
which are the default values, I get gitgutter sign rendered on top of ale sign, i.e. as if
it has higher priority than ale sign.
Anyway if I set:
```vim
let g:ale_sign_priority=30 
let g:gitgutter_sign_priority=9
```
i.e. a value less than 10 for gitgutter sign (where 10 is the default value assigned to signs by vim), the ale sign is correctly rendered with higher priority.
I suspect that ale_sign_priority is not handled correctly, i.e. the priority of ale signs is always set to the default value (i.e. 10). To support this conclusion, if set
 ```vim
let g:ale_sign_priority=8
let g:gitgutter_sign_priority=9
```
ale still has higher priority than gitgutter sign, being rendered on top of gitgutter sign.
In conclusion, it seems that ale signs are always rendered as if they have a priority equal to 10.

**Comment by benediktms** [comment]
Grettings,
I ran into the same issue with Syntastic, where gitgutter would get a higher priority in the sign column. the above fix of settign the following in the .vimrc seemed to fix this: 

`let g:syntastic_sign_priority=30`
`let g:gitgutter_sign_priority=9`

Just wanted to mention this here in case anyone else runs into the problem. Thanks for finding this :)
